README.md
# news organisations API

This API has the purpose of gathering an overview of existing news organizations by country and their RSS feeds (coming soon). This can be usefull as a base for gathering content for a news website/ application or for research purposes. This project is currently in devolpement and is as such not yet deployed.

## Getting started
### Run it locally
You will need [Docker](https://docs.docker.com/get-docker/) installed on your machine.
use `docker-compose up` in the root of the project to run the containers.

The API will be available on [127.0.0.1:3000](127.0.0.1:3000)

Please note that, except for a list of countries, the database will be empty. 
### Query the API

**Add a new news organization:**

POST /add-new-source 

with the body in the following format:
```
{
    name: "",
    country_id: "",
    website_url: ""
} 
```
**Get news organization by country:**

GET /sources/:country_id

**Get all news organizations:**

GET /get-all-publications

**Get all countries with their ID's:**

GET /get-all-countries

**Update news organization:**

GET /update-publication/:uuid

**Delete news organization by name:**

POST /delete-publication

with at least the following property in the body:
```
{
    name: ""
} 
```
**Delete country:**

GET /delete-country/:country_id

## I need help!

please contact me at axelle.vanden.eynde@student.ehb.be if google is unable to find your answer. 

## I want to contribute

Good! Please read the [contribution guidelines](https://github.com/axellevandeneynde/eindstuk-dev5/blob/main/CONTRIBUTING.md) to get yourself started.

## Tech stack
- server: [express.js](https://expressjs.com/)
- testing: [jest](https://jestjs.io/)
- database: [postgreSQL](https://www.postgresql.org/) whith [knex.js](http://knexjs.org/) for - queries.
- build: [Docker](https://docker.com)
- ci: [circleci](https://circleci.com/)
## Testing
all test can be found in `/api/src/__tests__`. You can run them by using the `npm run test` command in `/api`.




CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### v0.0.0

> 17 January 2021

- Endpoint delete country [`#9`](https://github.com/axellevandeneynde/eindstuk-dev5/pull/9)
- Endpoints get all publication & countries [`#8`](https://github.com/axellevandeneynde/eindstuk-dev5/pull/8)
- Endpoint update source [`#7`](https://github.com/axellevandeneynde/eindstuk-dev5/pull/7)
- Documentation [`#6`](https://github.com/axellevandeneynde/eindstuk-dev5/pull/6)
- Endpoint delete source by name [`#5`](https://github.com/axellevandeneynde/eindstuk-dev5/pull/5)
- Endpoint get sources from country [`#4`](https://github.com/axellevandeneynde/eindstuk-dev5/pull/4)
- Endpoint save new source [`#3`](https://github.com/axellevandeneynde/eindstuk-dev5/pull/3)
- Project setup [`#2`](https://github.com/axellevandeneynde/eindstuk-dev5/pull/2)
- ci: Add .circleci/config.yml [`#1`](https://github.com/axellevandeneynde/eindstuk-dev5/pull/1)
- build: adding & configure jest [`d8a6927`](https://github.com/axellevandeneynde/eindstuk-dev5/commit/d8a6927ba6c77480de252bddf0b3b1ed783bb0aa)
- feat(server.js): db connection & init tables [`92d453d`](https://github.com/axellevandeneynde/eindstuk-dev5/commit/92d453d694e2e6aa2754ff187a48fc52ae7e3401)
- refactor: change field names for countries table [`5f17d68`](https://github.com/axellevandeneynde/eindstuk-dev5/commit/5f17d682b3e70dfa7bf1d20c66a5e845e1e11254)


api/src/__tests__/helpers.test.js
const Helpers = require('../utils/helpers');

describe('tests to check isValidSourceObject()', () => {

    test('source has to be an object', () => {
        expect(Helpers.isValidSourceObject([])).toBe(false)
    });

    test('source should have the right propreties', () => {
        expect(Helpers.isValidSourceObject({ hey: 'you', bla: 'good' })).toBe(false)
    });

    test('countryId source country id should exist', () => {
        expect(Helpers.isValidSourceObject({
            name: 'de morgen', country_id: 'LAP', website_url: 'https://demorgen.be'
        })).toBe(false)
    });

    test('should return true when source is an object with right propreties and existing country id', () => {
        expect(Helpers.isValidSourceObject({
            name: 'de morgen', country_id: 'BE', website_url: 'https://demorgen.be'
        })).toBe(true)
    });
})

describe('test to check isValidPublicationNameObject()', () => {
    test('it has to be an object', () => {
        expect(Helpers.isValidPublicationNameObject([])).toBe(false)
    })
    test('it has to have a "name" property', () => {
        expect(Helpers.isValidPublicationNameObject({ bla: 'bla' })).toBe(false)
        expect(Helpers.isValidPublicationNameObject({ name: 'bla' })).toBe(true)
    })
})

describe('test to check isUuid', () => {
    test('has to be an uuid', () => {
        expect(Helpers.isUuid('78204893-2c9e-4fd8-a1bb-8786034ddb2a')).toBe(true)
        expect(Helpers.isUuid('something')).toBe(false)
    })
})

api/src/__tests__/end-to-end.test.js
const supertest = require('supertest');
const server = require('../server.js');

const request = supertest(server);

describe('end-to-end CRUD publication', () => {
    test('create publication', async () => {
        const res = await request
            .post('/add-new-source')
            .send({
                name: 'de morgen',
                country_id: 'BE',
                website_url: 'https://demorgen.be'
            });
        expect(await res.statusCode).toEqual(200)
    })

    test('read publication from the country', async () => {
        const res = await request
            .get('/sources/BE')
        expect(await res.statusCode).toEqual(200)
    })

    test('update publication with uuid', async () => {
        const res = await request
            .post(`/update-publication/78204893-2c9e-4fd8-a1bb-8786034ddb2a`).send({
                country_id: 'NL'
            })
        expect(await res.statusCode).toEqual(200)
    })

    test('delete publication with name', async () => {
        const res = await request
            .post('/delete-publication')
            .send({
                name: 'de morgen',
            });
        expect(await res.statusCode).toEqual(200)
        const data = await request.get('/sources/NL')
        let publicationWasDelted = true;
        data.body.publication.forEach(d => {
            if (d.name == 'de morgen') {
                publicationWasDelted = false;
            }
        })
        expect(publicationWasDelted).toBe(true)
    })
})

api/src/__tests__/integration/api-endpoints.test.js
const supertest = require('supertest');
const server = require('../../server.js')

const request = supertest(server);

describe('/add-new-source api endpoint', () => {
    test('should accept a post request with a new source object', async () => {
        const res = await request
            .post('/add-new-source')
            .send({
                name: 'de morgen',
                country_id: 'BE',
                website_url: 'https://demorgen.be'
            });
        expect(await res.statusCode).toEqual(200)
    });
    test('should not accept a post request with invalid body', async () => {
        const res = await request
            .post('/add-new-source')
            .send([{
                name: 'de morgen',
                country_id: 'BE',
                bla: 'https://demorgen.be'
            }]);
        expect(await res.statusCode).toEqual(400)
    });
})

describe('/sources/:country_id endpiont get sources from country', () => {
    test('should not accept non existing country ID', async () => {
        const res = await request
            .get('/sources/LAP')
        expect(await res.statusCode).toEqual(400)
    })
    test('should respond data if existing country ID', async () => {
        const res = await request
            .get('/sources/BE')
        expect(await res.body).toBeDefined()
    })
})

describe('/delete-publication with name', () => {
    test('should only accept delete request with name', async () => {
        const res = await request
            .post('/delete-publication').send({
                "bla": "bla"
            })
        expect(await res.statusCode).toEqual(400)
    })
    test('should respond to valid request positively', async () => {
        const res = await request
            .post('/delete-publication').send({
                "name": "The new york times"
            })
        expect(await res.body).toHaveProperty('message')
        expect(await res.statusCode).toEqual(200)
    })
})

describe('/update-publication with uuid', () => {
    test('should respond to valid request positively', async () => {
        const res = await request
            .post('/update-publication/78204893-2c9e-4fd8-a1bb-8786034ddb2a').send({
                name: 'the new york times'
            })
        expect(await res.body).toHaveProperty('message')
        expect(await res.statusCode).toEqual(200)
    })
    test('should respond to request without uuid negativly', async () => {
        const res = await request
            .post('/update-publication/something').send({
                name: 'the new york times'
            })
        expect(await res.statusCode).toEqual(400)
    })
    test('should respond to request with wrong proprety names negatively', async () => {
        const res = await request
            .post('/update-publication/something').send({
                post: 'the new york times'
            })
        expect(await res.statusCode).toEqual(400)
    })
    test('should respond to request with wrong proprety names negatively', async () => {
        const res = await request
            .post('/update-publication/78204893-2c9e-4fd8-a1bb-8786034ddb2a').send({
                post: 'the new york times'
            })
        expect(await res.statusCode).toEqual(400)
    })
    test('should respond to request with empty body negatively', async () => {
        const res = await request
            .post('/update-publication/78204893-2c9e-4fd8-a1bb-8786034ddb2a').send({})
        expect(await res.statusCode).toEqual(400)
    })
})

describe('/get-all-publications api endpoint', () => {
    test('should respond to valid request positively', async () => {
        const res = await request
            .get('/get-all-publications')
        expect(await res.statusCode).toEqual(200)
    })
})

describe('/get-all-countries api endpoint', () => {
    test('should respond to valid request positively', async () => {
        const res = await request
            .get('/get-all-countries')
        expect(await res.statusCode).toEqual(200)
    })
})

describe('/delete-country/:country_id endpoint', () => {
    test('should not accept non existing country ID', async () => {
        const res = await request
            .get('/delete-country/LAP')
        expect(await res.statusCode).toEqual(400)
    })
    test('should respond data if existing country ID', async () => {
        const res = await request
            .get('/delete-country/BE')
        expect(await res.statusCode).toEqual(200)
    })
})

server/jest.config.js
/*
 * For a detailed explanation regarding each configuration property, visit:
 * https://jestjs.io/docs/en/configuration.html
 */

module.exports = {
  // All imported modules in your tests should be mocked automatically
  // automock: false,

  // Stop running tests after `n` failures
  // bail: 0,

  // The directory where Jest should store its cached dependency information
  // cacheDirectory: "C:\\Users\\axell\\AppData\\Local\\Temp\\jest",

  // Automatically clear mock calls and instances between every test
  clearMocks: true,

  // Indicates whether the coverage information should be collected while executing the test
  // collectCoverage: false,

  // An array of glob patterns indicating a set of files for which coverage information should be collected
  // collectCoverageFrom: undefined,

  // The directory where Jest should output its coverage files
  coverageDirectory: "coverage",

  // An array of regexp pattern strings used to skip coverage collection
  coveragePathIgnorePatterns: [
    "\\\\node_modules\\\\"
  ],

  // Indicates which provider should be used to instrument code for coverage
  coverageProvider: "v8",

  // A list of reporter names that Jest uses when writing coverage reports
  // coverageReporters: [
  //   "json",
  //   "text",
  //   "lcov",
  //   "clover"
  // ],

  // An object that configures minimum threshold enforcement for coverage results
  // coverageThreshold: undefined,

  // A path to a custom dependency extractor
  // dependencyExtractor: undefined,

  // Make calling deprecated APIs throw helpful error messages
  // errorOnDeprecated: false,

  // Force coverage collection from ignored files using an array of glob patterns
  // forceCoverageMatch: [],

  // A path to a module which exports an async function that is triggered once before all test suites
  // globalSetup: undefined,

  // A path to a module which exports an async function that is triggered once after all test suites
  // globalTeardown: undefined,

  // A set of global variables that need to be available in all test environments
  // globals: {},

  // The maximum amount of workers used to run your tests. Can be specified as % or a number. E.g. maxWorkers: 10% will use 10% of your CPU amount + 1 as the maximum worker number. maxWorkers: 2 will use a maximum of 2 workers.
  // maxWorkers: "50%",

  // An array of directory names to be searched recursively up from the requiring module's location
  // moduleDirectories: [
  //   "node_modules"
  // ],

  // An array of file extensions your modules use
  // moduleFileExtensions: [
  //   "js",
  //   "json",
  //   "jsx",
  //   "ts",
  //   "tsx",
  //   "node"
  // ],

  // A map from regular expressions to module names or to arrays of module names that allow to stub out resources with a single module
  // moduleNameMapper: {},

  // An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
  // modulePathIgnorePatterns: [],

  // Activates notifications for test results
  // notify: false,

  // An enum that specifies notification mode. Requires { notify: true }
  // notifyMode: "failure-change",

  // A preset that is used as a base for Jest's configuration
  // preset: undefined,

  // Run tests from one or more projects
  // projects: undefined,

  // Use this configuration option to add custom reporters to Jest
  // reporters: undefined,

  // Automatically reset mock state between every test
  // resetMocks: false,

  // Reset the module registry before running each individual test
  // resetModules: false,

  // A path to a custom resolver
  // resolver: undefined,

  // Automatically restore mock state between every test
  // restoreMocks: false,

  // The root directory that Jest should scan for tests and modules within
  // rootDir: undefined,

  // A list of paths to directories that Jest should use to search for files in
  // roots: [
  //   "<rootDir>"
  // ],

  // Allows you to use a custom runner instead of Jest's default test runner
  // runner: "jest-runner",

  // The paths to modules that run some code to configure or set up the testing environment before each test
  // setupFiles: [],

  // A list of paths to modules that run some code to configure or set up the testing framework before each test
  // setupFilesAfterEnv: [],

  // The number of seconds after which a test is considered as slow and reported as such in the results.
  // slowTestThreshold: 5,

  // A list of paths to snapshot serializer modules Jest should use for snapshot testing
  // snapshotSerializers: [],

  // The test environment that will be used for testing
  testEnvironment: "node",

  // Options that will be passed to the testEnvironment
  // testEnvironmentOptions: {},

  // Adds a location field to test results
  // testLocationInResults: false,

  // The glob patterns Jest uses to detect test files
  // testMatch: [
  //   "**/__tests__/**/*.[jt]s?(x)",
  //   "**/?(*.)+(spec|test).[tj]s?(x)"
  // ],

  // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
  // testPathIgnorePatterns: [
  //   "\\\\node_modules\\\\"
  // ],

  // The regexp pattern or array of patterns that Jest uses to detect test files
  // testRegex: [],

  // This option allows the use of a custom results processor
  // testResultsProcessor: undefined,

  // This option allows use of a custom test runner
  // testRunner: "jasmine2",

  // This option sets the URL for the jsdom environment. It is reflected in properties such as location.href
  // testURL: "http://localhost",

  // Setting this value to "fake" allows the use of fake timers for functions such as "setTimeout"
  // timers: "real",

  // A map from regular expressions to paths to transformers
  // transform: undefined,

  // An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
  // transformIgnorePatterns: [
  //   "\\\\node_modules\\\\",
  //   "\\.pnp\\.[^\\\\]+$"
  // ],

  // An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
  // unmockedModulePathPatterns: undefined,

  // Indicates whether each individual test should be reported during the run
  // verbose: undefined,

  // An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
  // watchPathIgnorePatterns: [],

  // Whether to use watchman for file crawling
  // watchman: true,
};


api/src/server.js
const express = require('express');
const http = require('http');
const bodyParser = require('body-parser');
require('dotenv').config()
const Helpers = require('./utils/helpers')

//--------- Server setup --------
const app = express();
http.Server(app);

app.use(bodyParser.json());

//--------- DB CONNECTION --------
const pg = require('knex')({
    client: 'pg',
    version: '9.6',
    searchPath: ['knex', 'public'],
    connection: process.env.PG_CONNECTION_STRING ? process.env.PG_CONNECTION_STRING : 'postgres://example:example@localhost:5432/test'
});
// initialiseTables();


//--------- ROUTES --------

/**
* Endpoint (post) to create a new publication/source
* @params: {Object} publication - request body
* @params: {string} publication.name
* @params: {string} publication.country_id
* @params: {string} publication.website_url
* @returns: {object} message with uuid - response body
*/
app.post('/add-new-source', async (req, res) => {
    await initialiseTables();
    console.log('handling request /add-new-source');
    let newSource = req.body;
    if (Helpers.isValidSourceObject(newSource)) {
        newSource.uuid = Helpers.generateUUID();
        newSource.country_id = newSource.country_id.toUpperCase();
        pg.table('publications')
            .insert(newSource).then(
                res.status(200).send({ message: `news source was added with uuid ${newSource.uuid}` })
            )
    } else {
        res.status(400).send({ message: 'invalid news source object' })
    }

})


/**
* Endpoint (get) to get news publication/source with uuid
* @params: {string} uuid - request parameter
* @returns: {object} publication - response body
*/-
    app.get('/sources/:country_id', async (req, res) => {
        await initialiseTables();
        if (Helpers.isValidCountryId(req.params.country_id)) {
            pg.from('publications')
                .select(['name', 'website_url', 'country_id'])
                .where({ 'country_id': req.params.country_id.toUpperCase() })
                .then(data => {
                    res.status(200).send({ publication: data })
                })

        } else {
            res.status(400).send({ message: 'invalid country_id' })
        }
    })

/**
* Endpoint (post) to delete publication/source
* @params: {Object} publication - request body
* @params: {string} publication.name
* @params: {string} [publication.country_id]
* @params: {string} [publication.website_url]
* @returns: {object} message - response body
*/
app.post('/delete-publication', async (req, res) => {
    await initialiseTables();
    if (Helpers.isValidPublicationNameObject(req.body)) {
        await pg.from('publications').where({ name: req.body.name }).del();
        res.status(200).send({ message: 'delete request handled' })
    } else {
        res.status(400).send()
    }
})

/**
* Endpoint (post) to update a new publication/source
* @params: {string} uuid - request parameter
* @params: {Object} publication - request body
* @params: {string} publication.name
* @params: {string} publication.country_id
* @params: {string} publication.website_url
* @returns: {object} message with uuid - response body
*/
app.post('/update-publication/:uuid', async (req, res) => {
    await initialiseTables();
    console.log('handling request /update-publication')
    if (Helpers.isUuid(req.params.uuid) && Helpers.isValidPropertyNames(req.body)) {
        await pg.from('publications')
            .where({ uuid: req.params.uuid })
            .update(req.body)
        res.status(200).send({ message: 'update request handled' })
    } else {
        res.status(400).send()
    }
})

/**
* Endpoint (get) to get all publications joint with country
* @returns: {Object[]} publications - response body
*/
app.get('/get-all-publications', async (req, res) => {
    await initialiseTables();
    console.log('handling request /get-all-publications')
    await pg.select('*')
        .from('publications')
        .join('countries', 'countries.country_id', '=', 'publications.country_id')
        .then(data => {
            res.status(200).send(data);
        })

})

/**
* Endpoint (get) to get all countries
* @returns: {Object[]} countries - response body
*/
app.get('/get-all-countries', async (req, res) => {
    await initialiseTables();
    console.log('handling request /get-all-countries')
    await pg.from('countries')
        .select('*')
        .then(data => {
            res.status(200).send(data);
        })
})

/**
* Endpoint (get) to delete a country - DO NOT USE!
* @params: {string} country_id - request parameter
* @returns: {Object} message - response body
*/

// DO NOT USE!!! THIS IS JUST FOR EXERCICE PURPOSE.
app.get('/delete-country/:country_id', async (req, res) => {
    await initialiseTables();
    console.log('handling request /delete-all-countries')
    if (Helpers.isValidCountryId(req.params.country_id)) {
        await pg.from('publications').where({ country_id: req.params.country_id }).del();
        await pg.from('countries').where({ country_id: req.params.country_id }).del();
        res.status(200).send({ message: 'delete request handled' })
    } else {
        res.status(400).send({ message: 'invalid country_id' })
    }
})


//--------- INITIALISE TABLES --------

/**
* Checks if tables already exist in DB, 
* otherwhise tables are created and filled when applicable
*/
async function initialiseTables() {
    await pg.schema.hasTable("countries").then(async (exists) => {
        if (!exists) {
            await pg.schema
                .createTable("countries", (table) => {
                    // table.increments();
                    table.string("country");
                    table.string("country_id").unique().primary();
                })
                .then(async () => {
                    console.log("created table countries");
                    console.log("filling table countries...");
                    const countries = Helpers.getCountries();
                    countries.forEach(async (country) => {
                        await pg.table("countries").insert({ country: country.country, country_id: country.country_id });
                    });
                    console.log("table countries filled");
                })
        } else {
            console.log("table countries exists")
        }
    }).then(
        pg.schema.hasTable("publications").then(async (exists) => {
            if (!exists) {
                await pg.schema
                    .createTable("publications", (table) => {
                        table.uuid("uuid").unique().primary();
                        table.string("name");
                        table.string("website_url");
                        table.string("country_id").references("country_id").inTable("countries");
                    })
                    .then(async () => {
                        console.log("created table publications");
                    });
            } else {
                console.log('table publications exists')
            }
        }))
}

module.exports = app;